File: 39/39.py
from typing import List

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []

        def backtrack(start: int, path: List[int], remaining: int):
            if remaining == 0:
                res.append(list(path))
                return
            if remaining < 0:
                return
            
            for i in range(start, len(candidates)):
                path.append(candidates[i])  # 选择当前元素
                backtrack(i, path, remaining - candidates[i])  # 允许重复选择
                path.pop()  # 撤销选择

        backtrack(0, [], target)
        return res

File: 39/39.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> mid_vector;
        vector<vector<int>> result;
        solve(candidates, 0, 0, target, mid_vector, result);
        return result;
}

    void solve(vector<int>& candidates, int mid_target, int index, int target, vector<int>& mid_vector, vector<vector<int>>& result) {
        if (mid_target == target) {
            result.push_back(mid_vector);
            return;
        }
        if (mid_target > target) {
            return;
        }
        for (int i = index; i < candidates.size(); i++) {
            mid_vector.push_back(candidates[i]);
            //mid_target+=candidates[i];
            solve(candidates, mid_target + candidates[i], i, target, mid_vector, result);
            //solve(candidates, mid_target, i, target, mid_vector, result);
            mid_vector.pop_back();
        }
    }
};
